plutus-core/plutus-ir/src/PlutusIR/TypeCheck/Internal.hs

```diff
@@ -400,3 +401,16 @@ withTyVarsOfBindings = flip $ foldr withTyVarsOfBinding
 -- | Helper to add type variables into a computation's environment.
 withTyVarDecls :: [TyVarDecl TyName ann] -> TypeCheckM uni fun c e a -> TypeCheckM uni fun c e a
 withTyVarDecls = flip . foldr $ \(TyVarDecl _ n k) -> withTyVar n $ void k
+
+-- | Substitute `TypeBind`s from the given list of `Binding`s in the given `Type`.
+-- This is so that @let a = (con integer) in \(x : a) -> x@ typechecks.
+substTypeBinds ::
+    HasUniApply uni =>
+    NonEmpty (Binding TyName Name uni fun ann) ->
+    Normalized (Type TyName uni ()) ->
+    PirTCEnv uni fun e (Normalized (Type TyName uni ()))
+substTypeBinds = flip . foldrM $ \b ty -> case b of
+    TypeBind _ tvar rhs -> do
+        rhs' <- normalizeTypeM (void rhs)
+        substNormalizeTypeM rhs' (tvar ^. tyVarDeclName) (unNormalized ty)
```

Comment:
Ok, I think this is good w.r.t. global uniqueness. We're doing repeated calls of `substNormalizeTypeM`, so we only need to consider the base case of recursion (which is no concern for this PR) and the step, which preserves global uniqueness due to `substNormalizeTypeM` preserving it for the last argument even if its first one breaks it (i.e. we don't need to worry about `rhs`).